### PR TITLE
verify constants via array_key_exists

### PR DESCRIPTION
**Pull request by voku**
issue: https://github.com/phpstan/phpstan/issues/4174

I looked into ```ArrayKeyExistsFunctionTypeSpecifyingExtension```, but it seems like I didn't fully understand the logic of the specifying classes. In this case we want to specify parameter 1 or/and 2, can we currently do that?

### CODE REVIEW

**Comment by voku** [comment]
@voku reminder: try to to add two type specifying classes here, one for the first and the other for the second parameter

File: tests/PHPStan/Analyser/data/const-expr-phpdoc-type.php [hunk 1]
```diff
@@ -46,3 +46,121 @@ public function doFoo(
...
+	{
+		return [
+			self::NUMBER_TYPE_OFF      => 'Off',
+			self::NUMBER_TYPE_HEAD     => 'Head',
+			self::NUMBER_TYPE_POSITION => 'Position',
+		];
+	}
+
+	/**
+	 * @return array<int, HelloWorld::NUMBER_TYPE_*>
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/const-expr-phpdoc-type.php:76]: Phpdoc looks wrong

**Comment by voku** [comment]
@ondrejmirtes is there already a option to run multiple ```FunctionTypeSpecifyingExtension``` per function? I added a simple skip method, but I don't know if there is a better way.

**Comment by voku** [comment]
@ondrejmirtes Do you have some extra tests in mind?

File: src/Rules/Comparison/ImpossibleCheckTypeHelper.php [hunk 2]
```diff
@@ -132,6 +132,56 @@ public function findSpecifiedType(
 							}
 						}
 					}
+				} elseif (
+					$functionName === 'array_key_exists'
+					&& count($node->args) >= 2
+				) {
+					$haystackType = $scope->getType($node->args[1]->value);
+					if ($haystackType instanceof MixedType) {
```
Comment (staabm) [inline src/Rules/Comparison/ImpossibleCheckTypeHelper.php:140]: isn't mixed also covered by the following IF and therefore this case here can be dropped?

Comment (voku) [reply src/Rules/Comparison/ImpossibleCheckTypeHelper.php:140]: the same check is used for "in_array" here, so I think it's ok

**Comment by voku** [comment]
@ondrejmirtes this week I had again a bug in my code, where this fix could prevent it.

Can you maybe give feedback here, thanks. 😊

**Comment by herndlm** [comment]
I did not read through all comments, but I think we need to be careful that this doesn't become another `in_array` in regard of impossiblechecktype adaptions needed. Preferably none should be needed I guess :/

**Comment by ondrejmirtes** [comment]
Yes, that's what I worry about here. ImpossibleCheckTypeHelper should get thinner over time, not thicker.

**Comment by voku** [comment]
Maybe we can only use the "array_key_exists"-extension improvement? 🤔

**Comment by ondrejmirtes** [comment]
Hi, I'm cleaning up old and stale PRs. Please send a new PR if you're still interested, thanks.

I haven't merged this because I don't particularly feel great about the changes here.

**Pull request closed without merging**